db/external_sst_file_ingestion_job.cc

```diff
@@ -393,18 +412,48 @@ Status ExternalSstFileIngestionJob::Run() {
   }
   // It is safe to use this instead of LastAllocatedSequence since we are
   // the only active writer, and hence they are equal
-  SequenceNumber last_seqno = versions_->LastSequence();
+  SequenceNumber batch_start_last_seqno = versions_->LastSequence();
   edit_.SetColumnFamily(cfd_->GetID());
   // The levels that the files will be ingested into
 
-  for (IngestedFileInfo& f : files_to_ingest_) {
+  std::optional<int> prev_batch_uppermost_level;
+  for (auto& batch : file_batches_to_ingest_) {
+    SequenceNumber batch_end_last_seqno = 0;
+    int batch_uppermost_level = 0;
+    status = AssignLevelsForOneBatch(
+        batch, super_version, force_global_seqno, batch_start_last_seqno,
+        &batch_end_last_seqno, &batch_uppermost_level,
+        prev_batch_uppermost_level);
+    if (!status.ok()) {
+      return status;
+    }
+    batch_start_last_seqno = batch_end_last_seqno;
+    prev_batch_uppermost_level = batch_uppermost_level;
+  }
+
+  CreateEquivalentFileIngestingCompactions();
+  return status;
+}
+
+Status ExternalSstFileIngestionJob::AssignLevelsForOneBatch(
+    FileBatchInfo& batch, SuperVersion* super_version, bool force_global_seqno,
+    SequenceNumber batch_start_last_seqno, SequenceNumber* batch_end_last_seqno,
+    int* batch_uppermost_level, std::optional<int> prev_batch_uppermost_level) {
+  Status status;
+  assert(batch_uppermost_level);
+  *batch_uppermost_level = std::numeric_limits<int>::max();
+  // Make a copy of batch_start_last_seqno as we will be updating it during file
```

Comment:
This copy is not needed now?

Reply:
This copy is needed to keep updating `last_seqno` when iterating files, since we cannot directly updating `batch_start_last_seqno` and `AssignLevelAndSeqnoForIngestedFile` needs the last sequence number.

Reply:
> since we cannot directly updating batch_start_last_seqno

I don't see `batch_start_last_seqno` being used anywhere else in this function though.

Reply:
`batch_start_last_seqno` is just used to bootstrap `last_seqno`, it's not used anywhere else. But since `batch_start_last_seqno` is a function argument, we cannot directly use it, but need to make a copy of it to get the start sequence number, and then updating it in this block of code:

```   
if (assigned_seqno > last_seqno) {
      last_seqno = assigned_seqno;
      ++consumed_seqno_count_;
    }
``` 

 and then it will be passed to the next file's `AssignLevelAndSeqnoForIngestedFile` call.

Reply:
Yeah, you are right. This copy is not needed, I mistook the need for this copy with the actual fix for a previous bug. In fact, it's not needed to track sequence number usages within batches like this, I replaced it with one `last_seqno` that tracks sequence number across all batches.